### PR TITLE
Add new flag to allow element not to be cleaned up when open new file

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR is to fix defect [MAGN 7229 - Opening a new file while running auto deletes all Revit elements made with previous file in same Dynamo session](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7229). 

When a home workspace is being closed, the VM will try to garbage collect all objects that created by the .dyn file that associated with current home workspace. For objects that are from zero touch library, finally the VM will call `Dispose()` method on objects. But for some host application, for example Revit, all created objects shouldn't be cleaned up even workspace is closed. 

Previously, we used a global flag class `DisposeLogic` to indicate the state of Dynamo: it is being shutting down or not. But dispose logic really should be decided by host application. To decouple it from Dynamo, this PR made the following changes:
  1. Move `DisposeLogic` to Dynamo Revit.
  2. Add `WorkspaceRemoveStarted` event to `DynamoModel`, so for host application, if it inherits from `DynamoModel`, it could override `DynamoModel.OnWorkspaceRemoveStarted()` to get notification when a workspace is being removed. If the workspace is home workspace, `Dispose()` method will be invoked on objects after this event and the host application could use its own dispose logic to decide to really dispose or not. 

The fix on Revit side is [this PR](https://github.com/DynamoDS/DynamoRevit/pull/555).

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@ikeough do you have time to take a look?

### CODE REVIEW

File: src/DynamoCore/Models/WorkspaceModel.cs [hunk 1]
```diff
@@ -585,6 +585,7 @@ public virtual void Clear()
 
             foreach (NodeModel el in Nodes)
             {
+                el.RaisesModificationEvents = false;
```
Comment (ke-yu) [inline src/DynamoCore/Models/WorkspaceModel.cs]: Deleting connectors will trigger nodemodel's `OnModified` event, which consequently runs the graph with invalid inputs. Make nodes be silent.

**Comment by ikeough** [comment]
My gut instinct is that this is yet another flag (YAF :) ), and should be avoided. The way that we have to use this, setting it to true, making one method call, then setting it to false, is a bit unfortunate. But using the `DisposeLogic` class seems to be the only way to communicate to a zero touch method that something about the application state has changed.

**Comment by ke-yu** [comment]
@ikeough , you mean you track issue [MAGN-7229](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7229)?

File: src/DynamoCore/Models/WorkspaceModel.cs (hunk 1 above)
Comment (ikeough) [inline src/DynamoCore/Models/WorkspaceModel.cs]: Sorry. I meant that this line actually fixes something not related to this PR - that evaluation occurs when a workspace is cleared, but still really useful. Is there a YT task related to that issue?

**Comment by ke-yu** [comment]
@ikeough , adding event was my initial thought, but I thought this way is a little bit too heavy (because to other host application developers, if there is `HomeworkspaceClosing` event, they may expect more like `CustomworkspaceClosing`, `CustomworkspaceClosed`, `CustomworkspaceAdding`, `CustomworkspaceAdded` and so on).

The other way I tried was adding a virtual function `ClosingHomeworkspace()` to `DynamoModel` so that `DynamoRevitModel` could override it and get notification. 

But you are right, I don't like this flag neither. There are two issues: 1) Static class in most cases is horrible. 2) It doesn't work for multi homeworkspace cases. 

I'll try to implement with event.

**Comment by ikeough** [comment]
@ke-yu Thanks for the background.

I like your hybrid solution because it keeps the setting of `DisposeLogic.IsHomeWorkspaceClosing` to Revit model, which is the only place that currently uses it. I made some small suggestions around naming the pre-close event. We don't have other examples of this, but I vote for `WorkspaceRemoveStarted`.

**Comment by ke-yu** [comment]
@ikeough , yes, close and reopen .dyn file, all elements created by that .dyn file should still be reconnected back. 

I've updated the PR based on your comment. PTAL.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -1457,8 +1455,10 @@ public void RemoveWorkspace(WorkspaceModel workspace)
         {
             if (_workspaces.Remove(workspace))
             {
-                if (workspace is HomeWorkspaceModel)
+                OnWorkspaceRemoveStarted(workspace);
```
Comment (Randy-Ma) [inline src/DynamoCore/Models/DynamoModel.cs]: This is a bit like PreWorkspaceDisposed and I think it should be put just before the line where workspace is disposed.

File: src/NodeServices/WorkspaceEvents.cs [hunk 3]
```diff
@@ -19,6 +20,17 @@ public static void OnWorkspaceAdded(Guid id, string name)
         }
 
         /// <summary>
+        /// An event that is triggered when a workspace is being
+        /// removed from the DynamoModel's workspaces collection.
+        /// </summary>
+        public static event WorkspaceRemoveStartedEventHandler WorkspaceRemoveStarted;
+        public static void OnWorkspaceRemoveStarted(Guid id, string name)
+        {
+            if (WorkspaceRemoveStarted != null)
```
Comment (Randy-Ma) [inline src/NodeServices/WorkspaceEvents.cs]: It is better to call the event the same way as in DynamoModelEvents.OnWorkspaceRemoveStarted.

**Comment by Randy-Ma** [comment]
@ke-yu 
Could you help confirm these elements are in sync with the graph after the graph is closed and re-opened? For example, if you create 9 reference points(the number 9 is controlled by a slider), and close and re-open the graph, and then adjust the slider to create 6 points. Will the other 3 points be deleted correctly?

**Comment by Randy-Ma** [comment]
Sorry, it should have been deleted after reconciliation.

**Comment by ke-yu** [comment]
@Randy-Ma , the other 3 points be deleted correctly.

File: src/NodeServices/WorkspaceEvents.cs (hunk 3 above)
Comment (ikeough) [inline src/NodeServices/WorkspaceEvents.cs]: @ke-yu Can you also clarify in the comments when this event will actually be called? Something like:
```
 An event that is triggered prior to the removal of the workspace from the Workspaces collection.
```

**Comment by ke-yu** [comment]
Updated. Please take a look.

**Comment by ke-yu** [comment]
@Randy-Ma , you are write, after close and reopen, the elements are no longer associated with .dyn file.

**Comment by ikeough** [comment]
@ke-yu Sorry, I just saw that this was still open. I believe that, to satisfy the conditions of the YT task, the trace data should still exist for the call sites representing the nodes which were removed. Are you saying that the trace data is actually removed? If the dyn file was not saved, the trace data should still be there. 

Why are the elements no longer associated with the .dyn file?

**Comment by ke-yu** [comment]
@ikeough , I just verified that if .dyn file saved properly then the trace data does exist and next time open previous .dyn file, elements will be re-bind back to nodes. :-)

**Comment by ikeough** [comment]
LGTM. Thanks @ke-yu!

**Pull request merged**